### PR TITLE
Improving error messages for unsupported algorithms in Timestamp Provider

### PR DESCRIPTION
**Pull request by mishra14**
## Bug
Came up while discussing https://github.com/NuGet/Home/issues/6871
Regression: No

## Fix
Details: This PR improves the error messages for 2 scenarios in `Rfc3161TimestampProvider` - 

1. If the timestamping service certificate signature algorithm is not supported then the error message is improved from -
`The timestamp certificate has an unsupported signature algorithm.`
to - 
`The timestamp certificate has an unsupported signature algorithm - 'SHA1RSA'. The following algorithms are supported - 'SHA256RSA, SHA384RSA, SHA512RSA'.`

2. If the timestamping response digest algorithm is not supported then the error message is improved from -
`The timestamp certificate has an unsupported signature algorithm.`
to - 
`The timestamp response has an unsupported digest algorithm - 'SHA1'. The following algorithms are supported - 'SHA256, SHA384, SHA512'.`

## Testing/Validation
Tests Added: Yes

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/Strings.resx [hunk 1]
```diff
@@ -147,9 +147,14 @@
   </data>
   <data name="UnsupportedHashAlgorithm" xml:space="preserve">
     <value>Hash algorithm '{0}' is unsupported. Supported algorithms include: SHA512 and SHA256.</value>
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Common/Strings.resx:149]: should we also update this string to have the list of supported algorithms passed as an argument?

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs [hunk 2]
```diff
@@ -161,7 +161,17 @@ private static void ValidateTimestampCms(SigningSpecifications spec, SignedCms t
 
             if (!CertificateUtility.IsSignatureAlgorithmSupported(signerInfo.Certificate))
             {
-                throw new TimestampException(NuGetLogCode.NU3022, Strings.SignError_TimestampUnsupportedSignatureAlgorithm);
+                var certificateSignatureAlgorithm = signerInfo.Certificate.SignatureAlgorithm.FriendlyName?.ToUpper() ??
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs]: nit: extract this logic to a utility method to reuse it in both places yo do this.

**Review by PatoBeltran — APPROVED** [review]
🚀

File: src/NuGet.Core/NuGet.Common/Strings.resx (hunk 1 above)
Comment (mishra14) [reply src/NuGet.Core/NuGet.Common/Strings.resx:149]: The string is only used [here](https://github.com/NuGet/NuGet.Client/blob/7a8fe0bd7397e28a4fda6479b49e22df76578ece/src/NuGet.Core/NuGet.Common/CryptoHashProvider.cs#L50). The method only accepts SHA256 and SHA512.

File: src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs [hunk 3]
```diff
@@ -28,6 +28,21 @@ public sealed class SigningSpecificationsV1 : SigningSpecifications
         /// </summary>
         private static readonly string[] _allowedHashAlgorithmOids = _allowedHashAlgorithms.Select(hash => hash.ConvertToOidString()).ToArray();
 
+        /// <summary>
+        /// Allowed certifiacte signature algorithms.
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs]: Remove "certifiacte" --- you typo-ed it anyway.

File: src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs [hunk 4]
```diff
@@ -28,6 +28,21 @@ public sealed class SigningSpecificationsV1 : SigningSpecifications
...
+        /// </summary>
+        private static readonly SignatureAlgorithmName[] _allowedCertSignatureAlgorithms = new[]
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs]: Remove "Cert".

File: src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs [hunk 5]
```diff
@@ -28,6 +28,21 @@ public sealed class SigningSpecificationsV1 : SigningSpecifications
...
+        {
+            SignatureAlgorithmName.SHA256RSA,
+            SignatureAlgorithmName.SHA384RSA,
+            SignatureAlgorithmName.SHA512RSA
+        };
+
+        /// <summary>
+        /// Allowed certificate signature algorithm Oids.
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs]: Remove "certificate".

File: src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs [hunk 6]
```diff
@@ -28,6 +28,21 @@ public sealed class SigningSpecificationsV1 : SigningSpecifications
...
+        /// </summary>
+        private static readonly string[] _allowedCertSignatureAlgorithmOids = _allowedCertSignatureAlgorithms.Select(algorithm => algorithm.ConvertToOidString()).ToArray();
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs]: Remove "Cert"

File: src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs [hunk 7]
```diff
@@ -39,6 +54,10 @@ public sealed class SigningSpecificationsV1 : SigningSpecifications
 
         public override string[] AllowedHashAlgorithmOids => _allowedHashAlgorithmOids;
 
+        public override SignatureAlgorithmName[] AllowedCertificateSignatureAlgorithms => _allowedCertSignatureAlgorithms;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs]: Remove "Certificate".

File: src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs [hunk 8]
```diff
@@ -39,6 +54,10 @@ public sealed class SigningSpecificationsV1 : SigningSpecifications
...
+
+        public override string[] AllowedCertificateSignatureAlgorithmOids => _allowedCertSignatureAlgorithmOids;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Specifications/SigningSpecificationsV1.cs]: Remove "Certificate".

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs [hunk 9]
```diff
@@ -161,7 +161,16 @@ private static void ValidateTimestampCms(SigningSpecifications spec, SignedCms t
 
             if (!CertificateUtility.IsSignatureAlgorithmSupported(signerInfo.Certificate))
             {
-                throw new TimestampException(NuGetLogCode.NU3022, Strings.SignError_TimestampUnsupportedSignatureAlgorithm);
+                var certificateSignatureAlgorithm = GetNameOrOidString(signerInfo.Certificate.SignatureAlgorithm);
+
+                var supportedSignatureAlgorithms = string.Join(", ", spec.AllowedCertificateSignatureAlgorithms);
+
+                var errorMessage = string.Format(CultureInfo.CurrentCulture,
+                    Strings.SignError_TimestampCertificateUnsupportedSignatureAlgorithm,
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs]: I'd remove "SignError_" here and below.  This error should apply at signing time as well as verification time.  "SignError_" doesn't add much information.

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs [hunk 10]
```diff
@@ -196,7 +214,14 @@ private static void ValidateTimestampResponse(byte[] nonce, byte[] messageHash,
...
                 throw new TimestampException(NuGetLogCode.NU3019, Strings.SignError_TimestampIntegrityCheckFailed);
             }
+        }
 
+        /// <summary>
+        /// Returns the FriendlyName of an Oid. If FriendlyName is null, then the Oid string is returned.
+        /// </summary>
+        private static string GetNameOrOidString(Oid oid)
+        {
+            return oid.FriendlyName?.ToUpper() ?? oid.Value;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs:224]: Why force uppercase?

File: src/NuGet.Core/NuGet.Packaging/Strings.resx [hunk 11]
```diff
@@ -653,8 +650,10 @@ Valid from:</comment>
   <data name="SignError_TimestampSignatureValidationFailed" xml:space="preserve">
     <value>The timestamp signature validation failed.</value>
   </data>
-  <data name="SignError_TimestampUnsupportedSignatureAlgorithm" xml:space="preserve">
-    <value>The timestamp certificate has an unsupported signature algorithm.</value>
+  <data name="SignError_TimestampCertificateUnsupportedSignatureAlgorithm" xml:space="preserve">
+    <value>The timestamp certificate has an unsupported signature algorithm - '{0}'. The following algorithms are supported - '{1}'.</value>
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Strings.resx]: Don't use a hyphen like this, and you don't need single quotation marks.  Instead:

    The timestamp certificate has an unsupported signature algorithm ({0}). The following algorithms are supported:  {1}.

File: src/NuGet.Core/NuGet.Packaging/Strings.resx [hunk 12]
```diff
@@ -670,4 +669,12 @@ Valid from:</comment>
     <value>The file is not a valid nupkg. File path: {0}</value>
     <comment>0 - Nupkg file path</comment>
   </data>
+  <data name="ArgumentCannotBeNullOrEmpty" xml:space="preserve">
+    <value>The argument cannot be null or empty.</value>
+  </data>
+  <data name="SignError_TimestampResponseUnsupportedDigestAlgorithm" xml:space="preserve">
+    <value>The timestamp response has an unsupported digest algorithm - '{0}'. The following algorithms are supported - '{1}'.</value>
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Strings.resx]: The timestamp response has an unsupported digest algorithm ({0}). The following algorithms are supported:  {1}.

File: test/TestUtilities/Test.Utility/Signing/IssueCertificateOptions.cs [hunk 13]
```diff
@@ -69,6 +75,20 @@ public static IssueCertificateOptions CreateDefaultForEndCertificate()
             var id = CertificateUtilities.GenerateRandomId();
             var subjectName = new X509Name($"C=US,ST=WA,L=Redmond,O=NuGet,CN=NuGet Test Certificate ({id})");
 
+            return new IssueCertificateOptions()
+            {
+                KeyPair = keyPair,
+                SubjectName = subjectName,
+                SignatureAlgorithmName = "SHA256WITHRSA"
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/IssueCertificateOptions.cs]: Undo this change; it's already set by the default constructor.

File: test/TestUtilities/Test.Utility/Signing/IssueCertificateOptions.cs [hunk 14]
```diff
@@ -46,7 +50,8 @@ public static IssueCertificateOptions CreateDefaultForRootCertificateAuthority()
             {
                 KeyPair = keyPair,
                 IssuerPrivateKey = keyPair.Private,
-                SubjectName = subjectName
+                SubjectName = subjectName,
+                SignatureAlgorithmName = "SHA256WITHRSA"
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/IssueCertificateOptions.cs]: Undo this change; it's already set by the default constructor.

File: test/TestUtilities/Test.Utility/Signing/IssueCertificateOptions.cs [hunk 15]
```diff
@@ -59,7 +64,8 @@ public static IssueCertificateOptions CreateDefaultForIntermediateCertificateAut
             return new IssueCertificateOptions()
             {
                 KeyPair = keyPair,
-                SubjectName = subjectName
+                SubjectName = subjectName,
+                SignatureAlgorithmName = "SHA256WITHRSA"
```
Comment (dtivel) [inline test/TestUtilities/Test.Utility/Signing/IssueCertificateOptions.cs]: Undo this change; it's already set by the default constructor.

**Review by dtivel — CHANGES_REQUESTED** [review]

File: src/NuGet.Core/NuGet.Common/CryptoHashUtility.cs [hunk 16]
```diff
@@ -322,6 +325,27 @@ public static HashAlgorithmName OidToHashAlgorithmName(string oid)
             }
         }
 
+        /// <summary>
+        /// Extension method to convert NuGet.Common.SignatureAlgorithmName to an Oid string
+        /// </summary>
+        /// <returns>Oid string equivalent of the NuGet.Common.SignatureAlgorithmName</returns>
+        public static string ConvertToOidString(this SignatureAlgorithmName signatureAlgorithmName)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Common/CryptoHashUtility.cs:332]: I didn't see unit tests for this new method.  Please add them.

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs (hunk 10 above)
Comment (mishra14) [reply src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs:224]: To keep it uniform with our algorithm names, which are in upper case.

**Review by dtivel — APPROVED** [review]
Looks great.  Thanks!
